drivers/display/display_nrf_led_matrix.c

```diff
@@ -0,0 +1,448 @@
+	uint16_t pulse;
+
+	/* The timer is automagically stopped and cleared by shortcuts
+	 * on the same event (COMPARE0) that generates this interrupt.
+	 * But the event itself needs to be cleared here.
+	 */
+	nrf_timer_event_clear(dev_config->timer, NRF_TIMER_EVENT_COMPARE0);
+
+	/* Disable the row that contains the previously handled pixel. */
+	prev_row_idx = dev_config->pixel_mapping[dev_data->pixel_idx] >> 4;
+	set_pin(dev_config->rows[prev_row_idx], false);
+	/* Disconnect that pixel column pin from the peripheral driving it. */
+#if USE_PWM
+	nrf_pwm_disable(dev_config->pwm);
+#else
+	NRF_GPIOTE->CONFIG[dev_data->gpiote_ch] = 0;
+#endif
+
+	/* Switch to the next pixel. */
+	++dev_data->pixel_idx;
+	if (dev_data->pixel_idx >= PIXEL_COUNT) {
+		dev_data->pixel_idx = 0;
+	}
+	pixel_mapping = dev_config->pixel_mapping[dev_data->pixel_idx];
+	row_pin_info = dev_config->rows[pixel_mapping >> 4];
+	col_pin_info = dev_config->cols[pixel_mapping & 0xF];
+
+	/* Prepare the low pulse on the column pin for the current pixel. */
+	pulse = dev_data->framebuf[dev_data->pixel_idx] * QUANTUM;
+#if USE_PWM
+	dev_config->pwm->PSEL.OUT[0] = col_pin_info & PSEL_MASK;
+	dev_data->seq = pulse
+		      | ((col_pin_info & ACTIVE_LOW_MASK) ? 0 : BIT(15));
+	nrf_pwm_enable(dev_config->pwm);
+	nrf_pwm_task_trigger(dev_config->pwm, NRF_PWM_TASK_SEQSTART0);
+#else
+	uint32_t gpiote_cfg = GPIOTE_CONFIG_MODE_Task
+		| ((col_pin_info & PSEL_MASK) << GPIOTE_CONFIG_PSEL_Pos);
+
+	if (col_pin_info & ACTIVE_LOW_MASK) {
+		gpiote_cfg |= (GPIOTE_CONFIG_POLARITY_LoToHi
+			       << GPIOTE_CONFIG_POLARITY_Pos)
+			      /* If there should be no pulse at all for a given
+			       * pixel, its column GPIO needs to be configured
+			       * as initially inactive.
+			       */
+			   |  ((pulse == 0 ? GPIOTE_CONFIG_OUTINIT_High
+					   : GPIOTE_CONFIG_OUTINIT_Low)
+			       << GPIOTE_CONFIG_OUTINIT_Pos);
+	} else {
+		gpiote_cfg |= (GPIOTE_CONFIG_POLARITY_HiToLo
+			       << GPIOTE_CONFIG_POLARITY_Pos)
+			   |  ((pulse == 0 ? GPIOTE_CONFIG_OUTINIT_Low
+					   : GPIOTE_CONFIG_OUTINIT_High)
+			       << GPIOTE_CONFIG_OUTINIT_Pos);
+	}
+	nrf_timer_cc_set(dev_config->timer, 1, pulse);
+	NRF_GPIOTE->CONFIG[dev_data->gpiote_ch] = gpiote_cfg;
+#endif
+
+	/* Enable the row drive for the current pixel and restart the timer. */
+	set_pin(row_pin_info, true);
+	nrf_timer_task_trigger(dev_config->timer, NRF_TIMER_TASK_START);
+}
+
+static int instance_init(const struct device *dev)
+{
+	struct display_drv_data *dev_data = dev->data;
+	const struct display_drv_config *dev_config = dev->config;
+
+#if USE_PWM
+	uint32_t out_psels[NRF_PWM_CHANNEL_COUNT] = {
+		NRF_PWM_PIN_NOT_CONNECTED,
+		NRF_PWM_PIN_NOT_CONNECTED,
+		NRF_PWM_PIN_NOT_CONNECTED,
+		NRF_PWM_PIN_NOT_CONNECTED,
+	};
+	nrf_pwm_sequence_t sequence = {
+		.values.p_raw = &dev_data->seq,
+		.length = 1,
+	};
+
+	nrf_pwm_pins_set(dev_config->pwm, out_psels);
+	BUILD_ASSERT(PIXEL_PERIOD <= PWM_COUNTERTOP_COUNTERTOP_Msk);
```

Comment:
suggestion: I'd centralize build asserts in a single place for better readability